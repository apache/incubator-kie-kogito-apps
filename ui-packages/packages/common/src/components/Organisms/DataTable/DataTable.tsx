import React, { useEffect, useState } from 'react';
import { Bullseye } from '@patternfly/react-core';
import {
  Table,
  TableHeader,
  TableBody,
  ICell,
  IRow
} from '@patternfly/react-table';
import KogitoSpinner from '../../Atoms/KogitoSpinner/KogitoSpinner';
import {
  KogitoEmptyState,
  KogitoEmptyStateType
} from '../../Atoms/KogitoEmptyState/KogitoEmptyState';
import '@patternfly/patternfly/patternfly-addons.css';
import _ from 'lodash';
import uuidv4 from 'uuid';

interface IOwnProps {
  data: any[];
  isLoading: boolean;
  columns?: ICell[];
  networkStatus: any;
  error: any;
  refetch: () => void;
  LoadingComponent?: React.ReactNode;
  ErrorComponent?: React.ReactNode;
}

const getColumns = (data: any[], columns: ICell[]) => {
  let columnList: ICell[] = [];
  if (data) {
    columnList = columns
      ? _.filter(columns, column => !_.isEmpty(column.data))
      : _.filter(_.keys(_.sample(data)), key => key !== '__typename').map(
          key => ({ title: key, data: key } as ICell)
        );
  }
  return columnList;
};

const getRows = (data: any[], columns: ICell[]) => {
  let rowList: IRow[] = [];
  if (data) {
    rowList = data.map(rowData => {
      return {
        cells: _.reduce(
          columns,
          (result, column: ICell) => {
            _.forEach(rowData, (value, key) => {
              if (
                column.data &&
                key.toLowerCase() === column.data.toLowerCase()
              ) {
                if (_.isEmpty(value) || value === '{}') {
                  result.push('N/A');
                } else {
                  result.push(value);
                }
              }
            });
            return result;
          },
          []
        ),
        rowKey: uuidv4() // This is a walkaround to bypass the "id" cannot be included in "columns" issue
      };
    });
  }
  return rowList;
};

const DataTable: React.FC<IOwnProps> = ({
  data,
  isLoading,
  columns,
  networkStatus,
  error,
  LoadingComponent,
  ErrorComponent,
  refetch
}) => {
  const [rows, setRows] = useState<IRow[]>([]);
  const [columnList, setColumnList] = useState<ICell[]>([]);

  useEffect(() => {
    if (!_.isEmpty(data)) {
      setColumnList(getColumns(data, columns));
    }
  }, [data]);

  useEffect(() => {
    if (!_.isEmpty(data) && !_.isEmpty(columnList)) {
      setRows(getRows(data, columnList));
    }
  }, [columnList]);

  if (isLoading) {
    return LoadingComponent ? (
      <React.Fragment>{LoadingComponent}</React.Fragment>
    ) : (
      <Bullseye>
        <KogitoSpinner spinnerText="Loading..." />
      </Bullseye>
    );
  }

  if (networkStatus === 4) {
    return LoadingComponent ? (
      <React.Fragment>{LoadingComponent}</React.Fragment>
    ) : (
      <Bullseye>
        <KogitoSpinner spinnerText="Loading..." />
      </Bullseye>
    );
  }

  if (error) {
    return ErrorComponent ? (
      <React.Fragment>{ErrorComponent}</React.Fragment>
    ) : (
      <div className=".pf-u-my-xl">
        <KogitoEmptyState
          type={KogitoEmptyStateType.Refresh}
          title="Oops... error while loading"
          body="Try using the refresh action to reload user tasks"
          onClick={refetch}
        />
      </div>
    );
  }

  return (
    <React.Fragment>
      {data !== undefined &&
        !isLoading &&
        rows.length > 0 &&
        columnList.length > 0 && (
          <Table aria-label="Data Table" cells={columnList} rows={rows}>
            <TableHeader />
            <TableBody rowKey="rowKey" />
          </Table>
        )}
      {data !== undefined && !isLoading && rows.length === 0 && (
        <KogitoEmptyState
          type={KogitoEmptyStateType.Search}
          title="No results found"
          body="Try using different filters"
        />
      )}
    </React.Fragment>
  );
};

export default DataTable;
