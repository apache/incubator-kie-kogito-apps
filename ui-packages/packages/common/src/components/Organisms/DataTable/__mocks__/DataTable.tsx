import React from 'react';

const MockDataTable = (): React.ReactElement => {
  return <></>;
};

export default MockDataTable;
