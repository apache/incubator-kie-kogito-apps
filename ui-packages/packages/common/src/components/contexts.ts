import React from 'react';

export const aboutLogoContext = React.createContext('');
