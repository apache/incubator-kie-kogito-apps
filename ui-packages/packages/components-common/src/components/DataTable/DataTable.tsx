/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useEffect, useState } from 'react';
import { Bullseye } from '@patternfly/react-core';
import {
  Table,
  TableHeader,
  TableBody,
  IRow,
  ITransform,
  ICell,
  sortable,
  ISortBy
} from '@patternfly/react-table';
import '@patternfly/patternfly/patternfly-addons.css';
import _ from 'lodash';
import uuidv4 from 'uuid';
import jp from 'jsonpath';
import { OUIAProps, componentOuiaProps } from '../../utils/OuiaUtils';
import KogitoSpinner from '../KogitoSpinner/KogitoSpinner';
import {
  KogitoEmptyState,
  KogitoEmptyStateType
} from '../KogitoEmptyState/KogitoEmptyState';

export interface DataTableColumn {
  path: string;
  label: string;
  bodyCellTransformer?: (value: any, rowDataObj: Record<string, any>) => any;
  isSortable?: boolean;
}

interface IOwnProps {
  data: any[];
  isLoading: boolean;
  columns: DataTableColumn[];
  error: any;
  LoadingComponent?: React.ReactNode;
  ErrorComponent?: React.ReactNode;
  sortBy?: ISortBy;
  onSorting?: (index: number, direction: string) => void;
}

const getCellData = (dataObj: Record<string, unknown>, path: string) => {
  if (dataObj && path) {
    return !_.isEmpty(jp.value(dataObj, path))
      ? jp.value(dataObj, path)
      : 'N/A';
  } else {
    return 'N/A';
  }
};

const getColumns = (data: any[], columns: DataTableColumn[]) => {
  let columnList: ICell[] = [];
  if (data) {
    columnList = columns
      ? _.filter(columns, column => !_.isEmpty(column.path)).map(column => {
          return {
            title: column.label,
            data: column.path,
            cellTransforms: column.bodyCellTransformer
              ? [
                  ((value, extra) => {
                    const rowDataObj = data[extra.rowIndex];
                    return {
                      children: column.bodyCellTransformer(
                        value.title,
                        rowDataObj
                      )
                    };
                  }) as ITransform
                ]
              : undefined,
            transforms: column.isSortable ? [sortable] : undefined
          } as ICell;
        })
      : _.filter(_.keys(_.sample(data)), key => key !== '__typename').map(
          key => ({ title: key, data: `$.${key}` } as ICell)
        );
  }
  return columnList;
};

const getRows = (data: any[], columns: ICell[]) => {
  let rowList: IRow[] = [];
  if (data) {
    rowList = data.map(rowData => {
      return {
        cells: _.reduce(
          columns,
          (result, column: ICell) => {
            if (column.data) {
              result.push(getCellData(rowData, column.data));
            }
            return result;
          },
          []
        ),
        rowKey: uuidv4() // This is a walkaround to bypass the "id" cannot be included in "columns" issue
      };
    });
  }
  return rowList;
};

const DataTable: React.FC<IOwnProps & OUIAProps> = ({
  data,
  isLoading,
  columns,
  error,
  LoadingComponent,
  ErrorComponent,
  sortBy,
  onSorting,
  ouiaId,
  ouiaSafe
}) => {
  const [rows, setRows] = useState<IRow[]>([]);
  const [columnList, setColumnList] = useState<ICell[]>([]);

  useEffect(() => {
    if (data) {
      const cols = getColumns(data, columns);
      if (!_.isEmpty(cols)) {
        setColumnList(cols);
        setRows(getRows(data, cols));
      }
    }
  }, [data]);

  const onSort = (event, index, direction) => {
    if (_.isFunction(onSorting)) {
      onSorting(index, direction);
    }
  };

  if (isLoading) {
    return LoadingComponent ? (
      <React.Fragment>{LoadingComponent}</React.Fragment>
    ) : (
      <Bullseye>
        <KogitoSpinner spinnerText="Loading..." />
      </Bullseye>
    );
  }

  if (error) {
    return ErrorComponent ? (
      <React.Fragment>{ErrorComponent}</React.Fragment>
    ) : (
      <div className=".pf-u-my-xl">
        <KogitoEmptyState
          type={KogitoEmptyStateType.Refresh}
          title="Oops... error while loading"
          body="Try using the refresh action to reload"
        />
      </div>
    );
  }

  if (_.isEmpty(data)) {
    return (
      <KogitoEmptyState
        type={KogitoEmptyStateType.Search}
        title="No results found"
        body="Try using different filters"
      />
    );
  }

  return (
    <React.Fragment>
      <Table
        aria-label="Data Table"
        cells={columnList}
        rows={rows}
        sortBy={sortBy}
        onSort={onSort}
        {...componentOuiaProps(
          ouiaId,
          'data-table',
          ouiaSafe ? ouiaSafe : !isLoading
        )}
      >
        <TableHeader />
        <TableBody rowKey="rowKey" />
      </Table>
    </React.Fragment>
  );
};

export default DataTable;
