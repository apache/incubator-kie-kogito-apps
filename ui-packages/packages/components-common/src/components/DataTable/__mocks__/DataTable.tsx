import React from 'react';

const MockDataTable: React.FC = () => {
  return <></>;
};

export default MockDataTable;
