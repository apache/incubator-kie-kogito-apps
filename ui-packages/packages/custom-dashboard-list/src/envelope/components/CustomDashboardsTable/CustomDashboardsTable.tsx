/*
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useEffect, useState } from 'react';
import {
  DataTable,
  DataTableColumn,
  KogitoSpinner
} from '@kogito-apps/components-common';
import { OUIAProps, componentOuiaProps } from '@kogito-apps/ouia-tools';
import { CustomDashboardInfo, CustomDashboardListDriver } from '../../../api';
import {
  getDashboardNameColumn,
  getDateColumn
} from '../CustomDashboardListUtils/CustomDashboardListUtils';
import _ from 'lodash';
import { Bullseye } from '@patternfly/react-core';
import { ISortBy } from '@patternfly/react-table';

export interface CustomDashboardTableProps {
  driver: CustomDashboardListDriver;
  customDashboardData: CustomDashboardInfo[];
  setDashboardsData: React.Dispatch<
    React.SetStateAction<CustomDashboardInfo[]>
  >;
  isLoading: boolean;
}
interface SortBy {
  property: string;
  direction: 'asc' | 'desc';
}

const CustomDashboardsTable: React.FC<
  CustomDashboardTableProps & OUIAProps
> = ({
  driver,
  customDashboardData,
  setDashboardsData,
  isLoading,
  ouiaId,
  ouiaSafe
}) => {
  const [columns] = useState<DataTableColumn[]>([
    getDashboardNameColumn(
      (customDashboardInfo: CustomDashboardInfo): Promise<void> =>
        driver.openDashboard(customDashboardInfo)
    ),
    getDateColumn('lastModified', 'Last Modified')
  ]);
  const [sortBy, setSortBy] = useState<SortBy>({
    property: 'lastModified',
    direction: 'desc'
  });

  useEffect(() => {
    /* istanbul ignore else */
    if (!_.isEmpty(customDashboardData)) {
      onSort(2, 'desc');
    }
  }, [isLoading]);

  const getSortBy = (): ISortBy => {
    return {
      index: columns.findIndex((column) => column.path === sortBy.property),
      direction: sortBy.direction
    };
  };

  const onSort = async (index: number, direction): Promise<void> => {
    const sortObj: SortBy = {
      property: columns[index - 1].path,
      direction: direction.toLowerCase()
    };

    const sortedData = _.orderBy(
      customDashboardData,
      _.keys({
        [sortObj.property]: sortObj.direction
      }).map((key) => key),
      _.values({
        [sortObj.property]: sortObj.direction
      }).map((value) => value.toLowerCase())
    );
    setDashboardsData(sortedData);
    setSortBy(sortObj);
  };

  const customDashboardLoadingComponent: JSX.Element = (
    <Bullseye>
      <KogitoSpinner
        spinnerText="Loading Dashboard..."
        ouiaId="custom-dashboard-list-custom-dashboard-list"
      />
    </Bullseye>
  );

  return (
    <div {...componentOuiaProps(ouiaId, 'customDashboard-table', ouiaSafe)}>
      <DataTable
        data={customDashboardData}
        isLoading={isLoading}
        columns={columns}
        error={false}
        sortBy={getSortBy()}
        onSorting={onSort}
        LoadingComponent={customDashboardLoadingComponent}
      />
    </div>
  );
};

export default CustomDashboardsTable;
