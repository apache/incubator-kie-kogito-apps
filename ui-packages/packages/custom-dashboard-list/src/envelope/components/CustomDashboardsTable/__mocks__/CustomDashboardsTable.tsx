import React from 'react';

const MockedCustomDashboardTable = (): React.ReactElement => {
  return <></>;
};

export default MockedCustomDashboardTable;
