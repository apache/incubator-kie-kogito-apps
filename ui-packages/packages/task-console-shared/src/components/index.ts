export { default as TaskState } from './TaskState/TaskState';
