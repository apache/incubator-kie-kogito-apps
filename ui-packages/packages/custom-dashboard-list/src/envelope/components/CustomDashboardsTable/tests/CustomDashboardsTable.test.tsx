/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from 'react';
import { render, screen } from '@testing-library/react';
import CustomDashboardsTable from '../CustomDashboardsTable';
import {
  customDashboardInfos,
  MockedCustomDashboardListDriver
} from '../../../tests/mocks/MockedCustomDashboardsListDriver';
import { DataTable } from '@kogito-apps/components-common/dist/components/DataTable';

Date.now = jest.fn(() => 1487076708000); //14.02.2017

describe('customDashboard table test', () => {
  const driver = new MockedCustomDashboardListDriver();
  it('renders table', () => {
    const { container } = render(
      <CustomDashboardsTable
        driver={driver}
        isLoading={false}
        customDashboardData={customDashboardInfos}
        setDashboardsData={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
    const checkTable = container.querySelector('table');
    expect(checkTable).toBeTruthy();
  });
});
