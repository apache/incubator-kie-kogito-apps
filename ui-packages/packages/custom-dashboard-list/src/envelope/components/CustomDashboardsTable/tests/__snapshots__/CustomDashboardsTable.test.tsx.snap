// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customDashboard table test renders table 1`] = `
<div>
  <div
    data-ouia-component-type="customDashboard-table"
    data-ouia-safe="true"
  >
    <table
      aria-label="Data Table"
      class="pf-c-table pf-m-grid-md"
      data-ouia-component-id="OUIA-Generated-Table-2"
      data-ouia-component-type="data-table"
      data-ouia-safe="true"
      role="grid"
    >
      <thead
        class=""
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-1"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <th
            aria-sort="none"
            class="pf-c-table__sort"
            data-key="0"
            data-label="Name"
            scope="col"
          >
            <button
              class="pf-c-table__button"
              type="button"
            >
              <div
                class="pf-c-table__button-content"
              >
                <span
                  class="pf-c-table__text"
                >
                  Name
                </span>
                <span
                  class="pf-c-table__sort-indicator"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style="vertical-align: -0.125em;"
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                    />
                  </svg>
                </span>
              </div>
            </button>
          </th>
          <th
            aria-sort="descending"
            class="pf-c-table__sort pf-m-selected"
            data-key="1"
            data-label="Last Modified"
            scope="col"
          >
            <button
              class="pf-c-table__button"
              type="button"
            >
              <div
                class="pf-c-table__button-content"
              >
                <span
                  class="pf-c-table__text"
                >
                  Last Modified
                </span>
                <span
                  class="pf-c-table__sort-indicator"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style="vertical-align: -0.125em;"
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M168 345.941V44c0-6.627-5.373-12-12-12h-56c-6.627 0-12 5.373-12 12v301.941H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.569 9.373 33.941 0l86.059-86.059c15.119-15.119 4.411-40.971-16.971-40.971H168z"
                    />
                  </svg>
                </span>
              </div>
            </button>
          </th>
        </tr>
      </thead>
      <tbody
        class=""
        role="rowgroup"
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-2"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-key="0"
            data-label="Name"
          >
            <a>
              <strong>
                dashboard1
              </strong>
            </a>
          </td>
          <td
            class=""
            data-key="1"
            data-label="Last Modified"
          >
            <time
              datetime="NaN"
            >
              Invalid date
            </time>
          </td>
        </tr>
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-3"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-key="0"
            data-label="Name"
          >
            <a>
              <strong>
                dashboard2
              </strong>
            </a>
          </td>
          <td
            class=""
            data-key="1"
            data-label="Last Modified"
          >
            <time
              datetime="NaN"
            >
              Invalid date
            </time>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
